boards: Add support for Wio Terminal (#54953)
zephyrproject-rtos/zephyr: boards/arm/wio_terminal/doc/index.rst

| GPIO      | on-chip    | I/O ports                                |
+-----------+------------+------------------------------------------+
| USART     | on-chip    | Serial port                              |
+-----------+------------+------------------------------------------+
| I2C       | on-chip    | Inter-Integrated Circuit                 |
+-----------+------------+------------------------------------------+
| SPI       | on-chip    | Serial Peripheral Interface port         |
+-----------+------------+------------------------------------------+
| TRNG      | on-chip    | True Random Number Generator             |
+-----------+------------+------------------------------------------+
| HWINFO    | on-chip    | Unique 128 bit serial number             |
+-----------+------------+------------------------------------------+
| RTC       | on-chip    | Real-Time Counter                        |
+-----------+------------+------------------------------------------+
| USB       | on-chip    | USB device                               |
+-----------+------------+------------------------------------------+
| PWM       | on-chip    | PWM                                      |

> +-----------+------------+------------------------------------------+

Reviewer: Update table to new format, see [atsamr34_xpro](https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/arm/atsamr34_xpro/doc/index.rst) as an example.

Author: Done

Other hardware features are not currently supported by Zephyr.

The default configuration can be found in the Kconfig file
:zephyr_file:`boards/arm/wio_terminal/wio_terminal_defconfig`.

Zephyr can use the default Cortex-M SYSTICK timer or the SAM0 specific RTC.
To use the RTC, set :kconfig:option:`CONFIG_CORTEX_M_SYSTICK=n` and set
:kconfig:option:`CONFIG_SYS_CLOCK_TICKS_PER_SEC` to no more than 32 kHZ
divided by 7, i.e. no more than 4500.

Connections and IOs
===================

The `Wio Terminal Getting started guide`_ has detailed
information about the board including `pinouts`_ and its `schematics`_.

System Clock
============

The SAMD51 MCU is configured to use the 32.768 kHz internal oscillator
with the on-chip PLL generating the 120 MHz system clock.

Serial Port
===========

The SAMD51 MCU has 8 SERCOM based USARTs. On the Wio Terminal, no SERCOM is
available for the Zephyr console and indeed it is configured by default over

> the USB interface.

Reviewer: Console will be by USB, require update this section.

Author: Done

.. _wio_terminal:

Wio Terminal
############

Overview
********

The Wio Terminal is a small (72 mm x 57 mm x 12 mm) and powerful ARM board with

> wireless connectivity (2.4G/5G Dual Bands Wireless and BLE5.0), LCD display,

Reviewer 2: ```suggestion
wireless connectivity (2.4G/5G dual-band Wi-Fi and BLE 5.0), LCD display,
```

> USB C port, FPC connector, MicroSD card slot, Raspberry pi compatible 40pin

Reviewer 2: ```suggestion
USB-C port, FPC connector, microSD card slot, Raspberry Pi compatible 40-pin
```

To use the RTC, set :kconfig:option:`CONFIG_CORTEX_M_SYSTICK=n` and set
:kconfig:option:`CONFIG_SYS_CLOCK_TICKS_PER_SEC` to no more than 32 kHZ divided
by 7, i.e. no more than 4500.

Connections and IOs
===================

The `Wio Terminal Getting started guide`_ has detailed information about the
board including `pinouts`_ and its `schematics`_.

System Clock
============

The SAMD51 MCU is configured to use the 32.768 kHz internal oscillator with the
on-chip PLL generating the 120 MHz system clock.

Serial Port
===========

Zephyr console output is available using the USB connector, which is used to
make the console available on PC as USB CDC class.

USB Device Port
===============

The SAMD51 MCU has a USB device port that can be used to communicate with a
host PC.  See the :ref:`usb-samples` sample applications for more, such as the
:ref:`usb_cdc-acm` sample which sets up a virtual serial port that echos
characters back to the host PC.

Programming and Debugging
*************************

The Wio Terminal ships with a the BOSSA compatible UF2 bootloader. The
bootloader can be entered by quickly tapping the reset button twice.

> The UF2 file is generated when building, it is possible to use it to flash the

Reviewer 2: @joelguittet sorry, look like I should have suggested the change differently, as it was more than one line. Could you please incorporate the full change, as it fixes improper grammar issues in the rest of the paragraph?

target as much as simple as copy/paste the UF2 to the drive mounted when
entering the bootloader. The device resets on the new firmware after the UF2
file is loaded.

Flashing
========

#. Build the Zephyr kernel and the :code:`button` sample application:

   .. zephyr-app-commands::
      :zephyr-app: samples/basic/button
      :board: wio_terminal
      :goals: build
      :compact:

#. Swipe the reset/power button down twice quickly to enter bootloader mode

#. Flash the image:

   .. zephyr-app-commands::
      :zephyr-app: samples/basic/button
      :board: wio_terminal
      :goals: flash
      :compact:

   You should see the blue (user) LED flashing whenever you press the third
   (counting from the top left) user button at the top of the Wio Terminal.

Debugging
=========

In addition to the built-in bootloader, the Wio Terminal can be flashed and

> debugged using a SWD probe such as the Segger J-Link.

Reviewer 2: ```suggestion
debugged using an SWD probe such as the Segger J-Link.
```

.. _wio_terminal:

Wio Terminal
############

Overview
********

The Wio Terminal is a small (72 mm x 57 mm x 12 mm) and powerful ARM board
with wireless connectivity (2.4G/5G Dual Bands Wireless and BLE5.0), LCD
display, USB C port, FPC connector, MicroSD card slot, Raspberry pi compatible
40pin header and 2 Grove connectors.

.. image:: img/wio_terminal.png
     :width: 500px
     :align: center

>      :alt: Wio Terminal

Reviewer 2: ```suggestion
     :alt: Seeed Studio Wio Terminal
```

>      :alt: Wio Terminal

Author: The name is "Wio Terminal", not "Seed Studio Terminal". I disagree with the suggestion. I propose instead "Seed Studio Wio terminal" if this is suitable.

>      :alt: Wio Terminal

Reviewer 2: Yes sorry that was a typo on my side :) and it's Seeed, not Seed!

>      :alt: Wio Terminal

Author: So we agree for "Seeed Studio Wio Terminal", correct ?

Hardware
********

> - ATSAMD51P19 ARM Cortex-M4 processor at 120 MHz

Reviewer 2: ```suggestion
- ATSAMD51P19 ARM Cortex-M4F processor at 120 MHz (boost up to 200 MHz)
```

> - ATSAMD51P19 ARM Cortex-M4 processor at 120 MHz

Author: Done

- 512 KiB flash memory and 192 KiB of RAM
- 4 MiB external flash
- MicroSD card slot
- RTL8720DN 2.4G/5G Dual Bands Wireless and BLE5.0 Combo Module
- 2.4inch LCD display
- LIS3DH accelerometer
- Microphone 1.0V-10V -42dB
- Speaker ≥78dB @10cm 4000Hz
- Light Sensor 400-1050nm
- Infrared Emitter 940nm
- GPIO 40 pin (Raspberry Pi compatible)
- 2x Grove connectors
- 1x user LED
- 3x user buttons
- 5-way user button

> - power/reset/boot mode switch

Reviewer 2: ```suggestion
- Power/Reset/Boot mode switch
```

> - power/reset/boot mode switch

Author: Done

+-----------+------------+------------------------------------------+
| SPI       | on-chip    | Serial Peripheral Interface port         |
+-----------+------------+------------------------------------------+
| TRNG      | on-chip    | True Random Number Generator             |
+-----------+------------+------------------------------------------+
| HWINFO    | on-chip    | Unique 128 bit serial number             |
+-----------+------------+------------------------------------------+
| RTC       | on-chip    | Real-Time Counter                        |
+-----------+------------+------------------------------------------+
| USB       | on-chip    | USB device                               |
+-----------+------------+------------------------------------------+
| PWM       | on-chip    | PWM                                      |
+-----------+------------+------------------------------------------+

Other hardware features are not currently supported by Zephyr.

The default configuration can be found in the Kconfig file
:zephyr_file:`boards/arm/wio_terminal/wio_terminal_defconfig`.

Zephyr can use the default Cortex-M SYSTICK timer or the SAM0 specific RTC.

> To use the RTC, set :code:`CONFIG_CORTEX_M_SYSTICK=n` and set

Author: Done

> :code:`CONFIG_SYS_CLOCK_TICKS_PER_SEC` to no more than 32 kHZ divided by 7,

Author: Done

i.e. no more than 4500.

Connections and IOs
===================

The `Wio Terminal Getting started guide`_ has detailed

> information about the board including `pinouts`_ and the `schematic`_.

Reviewer 2: ```suggestion
information about the board including `pinouts`_ and its `schematics`_.
```

> information about the board including `pinouts`_ and the `schematic`_.

Author: Done

System Clock
============

The SAMD51 MCU is configured to use the 32.768 kHz internal oscillator
with the on-chip PLL generating the 120 MHz system clock.

Serial Port
===========

The SAMD51 MCU has 8 SERCOM based USARTs. On the Wio Terminal, no SERCOM
is available for the Zephyr console but it is possible to use SERCOM2
available on Raspberry Pi compatible header pins 8 (TX) and 10 (RX).

To add Zephyr console over Raspberry Pi connector, add the following
configuration and overlay in your project:

```
CONFIG_CONSOLE=y
CONFIG_UART_CONSOLE=y
CONFIG_SERIAL=y
```

```
/ {
    chosen {

>         zephyr,console = &raspberrypi_serial;

Reviewer 2: Why not already put it in wio_terminal.dts by default?

>         zephyr,console = &raspberrypi_serial;

Author: Because there is no real console port on the Wio Terminal. This is a serial port available on the Raspberry Pi connector. Not even very easy to connect an FTDI or something similar without a bit of wiring.
So best choice is no console by default. I have added here this comment because many people will look for a console. This is a solution, even if the USB console is very nice too.

        zephyr,shell-uart = &raspberrypi_serial;
    };
};
```

Note it is also possible to use USB CDC interface to get a Zephyr console
over USB. See the :ref:`samples/subsys/usb/console` to have a working example.

USB Device Port
===============

The SAMD51 MCU has a USB device port that can be used to communicate
with a host PC.  See the :ref:`usb-samples` sample applications for
more, such as the :ref:`usb_cdc-acm` sample which sets up a virtual
serial port that echos characters back to the host PC.

Programming and Debugging
*************************

The Wio Terminal ships with a the BOSSA compatible UF2 bootloader. The
bootloader can be entered by quickly tapping the reset button twice.

> Additionally, if :code:`CONFIG_USB_CDC_ACM` is enabled then the bootloader

Author: Done

   Wio Terminal.

Debugging
=========

In addition to the built-in bootloader, the Wio Terminal can be flashed and
debugged using a SWD probe such as the Segger J-Link.

#.  Solder cables to the code:`SWCLK`, :code:`SWDIO`, :code:`RESET`,
    :code:`GND`, and :code:`3V3` pins. See `Test with SWD`_ for more
    information.

#. Connect the board to the probe by connecting the :code:`SWCLK`,
   :code:`SWDIO`, :code:`RESET`, :code:`GND`, and :code:`3V3` pins on the
   Wio Terminal to the :code:`SWCLK`, :code:`SWDIO`, :code:`RESET`,
   :code:`GND`, and :code:`VTref` pins on the `J-Link`_.

#. Flash the image:

   .. zephyr-app-commands::
      :zephyr-app: samples/basic/button
      :board: wio_terminal
      :goals: flash -r openocd
      :compact:

#. Start debugging:

   .. zephyr-app-commands::
      :zephyr-app: samples/basic/button
      :board: wio_terminal
      :goals: debug
      :compact:

References
**********

.. target-notes::

.. _Wio Terminal Getting started guide:
   https://wiki.seeedstudio.com/Wio-Terminal-Getting-Started/

.. _pinouts:
    https://wiki.seeedstudio.com/Wio-Terminal-Getting-Started/#pinout-diagram

> .. _schematic:

Reviewer 2: ```suggestion
.. _schematics:
```

> .. _schematic:

Author: Done
